src/core/server/saved_objects/migrations/core/migration_context.ts

```diff
@@ -107,20 +111,68 @@ function createSourceContext(source: FullIndexInfo, alias: string) {
 function createDestContext(
   source: FullIndexInfo,
   alias: string,
-  mappingProperties: SavedObjectsTypeMappingDefinitions
+  typeMappingDefinitions: SavedObjectsTypeMappingDefinitions
 ): FullIndexInfo {
-  const activeMappings = buildActiveMappings(mappingProperties);
+  const targetMappings = disableUnknownTypeMappingFields(
+    buildActiveMappings(typeMappingDefinitions),
+    source.mappings
+  );
 
   return {
     aliases: {},
     exists: false,
     indexName: nextIndexName(source.indexName, alias),
-    mappings: {
-      ...activeMappings,
-      properties: {
-        ...source.mappings.properties,
-        ...activeMappings.properties,
-      },
+    mappings: targetMappings,
+  };
+}
+
+/**
+ * Merges the active mappings and the source mappings while disabling the
+ * fields of any unknown Saved Object types present in the source index's
+ * mappings.
+ *
+ * Since the Saved Objects index has `dynamic: strict` defined at the
+ * top-level, only Saved Object types for which a mapping exists can be
+ * inserted into the index. To ensure that we can continue to store Saved
+ * Object documents belonging to a disabled plugin we define a mapping for all
+ * the unknown Saved Object types that were present in the source index's
+ * mappings. To limit the field count as much as possible, these unkwnown
+ * type's mappings are set to `dynamic: false`.
+ *
+ * (Since we're using the source index mappings instead of looking at actual
+ * document types in the inedx, we potentially add more "unknown types" than
+ * what would be necessary to support migrating all the data over to the
+ * target index.)
+ *
+ * @param activeMappings The mappings compiled from all the Saved Object types
+ * known to this Kibana node.
+ * @param sourceMappings The mappings of index used as the migration source.
+ * @returns The mappings that should be applied to the target index.
+ */
+function disableUnknownTypeMappingFields(
+  activeMappings: IndexMapping,
+  sourceMappings: IndexMapping
+): IndexMapping {
+  const targetTypes = Object.keys(activeMappings.properties);
+
+  const disabledTypesProperties = Object.keys(sourceMappings.properties)
+    .filter((sourceType) => {
+      const isObjectType = 'properties' in sourceMappings.properties[sourceType];
+      // Only Object/Nested datatypes can be excluded from the field count by
+      // using `dynamic: false`.
+      return !targetTypes.includes(sourceType) && isObjectType;
+    })
+    .reduce((disabledTypesAcc, sourceType) => {
+      disabledTypesAcc[sourceType] = { dynamic: false, properties: {} };
+      return disabledTypesAcc;
+    }, {} as SavedObjectsMappingProperties);
```

Comment:
May be a stupid question but: Can we actually go deeper, by performing recursion within the fields that actually got nested fields (properties)?

Reply:
Every level adds a field so a mapping like:
```js
{
  "dynamic": false,
  "properties": {
    "a": {
      "properties": {
        "b": {
          "properties": {
            "c": {
              "type": "text"
            }
          }
        }
      }
    }
  }
}
```

Would add fields `a`, `b` and `c` that count towards the field count limit (even though only `c` is added to the index).

Reply:
So the lowest field count is actually if we drop the unknown field completely from mappings and use `dynamic: false` at the top-level (the current PR still has one field remaining for every unknown type). But that will mean we loose any validation that ES used to provide and we will have to expose validation from the saved objects "framework" so plugins can add the validation that ES used to provide.

Reply:
I've fleshed out the PR description a bit better to document this, since I think it's important to know that this doesn't completely solve the problem.